ext/drbbdup/drbbdup.c

```diff
@@ -697,25 +706,35 @@ drbbdup_encode_runtime_case(void *drcontext, drbbdup_per_thread *pt, void *tag,
     }
 
     /* Encoding is application-specific and therefore we need to user to define the
-     * encoding of the runtime case. We invoke a user-defined call-back.
-     */
-    ASSERT(opts.insert_encode, "The encode call-back cannot be NULL");
-    /* Note, we could tell the user not to reserve flags and scratch register since
-     * drbbdup is doing that already. However, for flexibility/backwards compatibility
-     * ease, this might not be the best approach.
+     * encoding of the runtime case. Therefore, we invoke a user-defined call-back.
+     *
+     * It could also be the encoding is done directly and changed on demand. Therefore,
+     * the call-back may be NULL.
      */
-    opts.insert_encode(drcontext, tag, bb, where, opts.user_data, pt->orig_analysis_data);
+    if (opts.insert_encode != NULL) {
+        /* Note, we could tell the user not to reserve flags and scratch register since
+         * drbbdup is doing that already. However, for flexibility/backwards compatibility
+         * ease, this might not be the best approach.
+         */
+        opts.insert_encode(drcontext, tag, bb, where, opts.user_data,
+                           pt->orig_analysis_data);
 
-    /* Restore all unreserved registers used by the call-back. */
-    drreg_restore_all(drcontext, bb, where);
+        /* Restore all unreserved registers used by the call-back. */
+        drreg_restore_all(drcontext, bb, where);
+    }
 
 #ifdef X86_32
     /* Load the encoding to the scratch register.
      * The dispatcher could compare directly via mem, but this will
      * destroy micro-fusing (mem and immed).
      */
     opnd_t scratch_reg_opnd = opnd_create_reg(DRBBDUP_SCRATCH_REG);
-    opnd_t opnd = drbbdup_get_encoding_opnd();
+    opnd_t opnd;
+#    ifdef DEBUG
+    drbbdup_status_t res =
```

Comment:
Use `IF_DEBUG(drbbdup_status_t res =)` for more readable code.  Ditto below.

Reply:
Done.